File: getstatus_lambda/lambda_function.py
import base64
import boto3
import os
import json
import logging

logger = logging.getLogger()
logger.setLevel(logging.INFO)

def lambda_handler(event, context):
    """
    Lambda function to retrieve the latest scheduled job status for a user.
    Queries DynamoDB for the most recent item by user_id, ordered by added_at descending.
    Returns schedule_time and status.
    """
    try:
        logger.info(f"Received event: {json.dumps(event)}")
        # Extract user_id from event (assuming it's passed in the body or directly)
        if 'body' in event:
            if event.get('isBase64Encoded'):
                body = json.loads(base64.b64decode(event['body']).decode())
            else:
                body = json.loads(event['body'])
            user_id = body.get('user_id')
        else:
            user_id = event.get('user_id')

        if not user_id:
            return {
                'statusCode': 400,
                'body': json.dumps({'error': 'user_id is required'})
            }

        # Get DynamoDB table
        region = os.environ.get('AWS_APP_REGION', 'eu-north-1')
        table_name = os.environ.get('DYNAMODB_TABLE')
        if not table_name:
            return {
                'statusCode': 500,
                'body': json.dumps({'error': 'DYNAMODB_TABLE environment variable not set'})
            }

        dynamodb = boto3.resource('dynamodb', region_name=region)
        table = dynamodb.Table(table_name)

        # Query for the most recent item by user_id (assuming added_at is sort key)
        response = table.query(
            KeyConditionExpression=boto3.dynamodb.conditions.Key('user_id').eq(user_id),
            ScanIndexForward=False,  # Descending order to get most recent
            Limit=1
        )

        if not response['Items']:
            return {
                'statusCode': 404,
                'body': json.dumps({'error': 'No scheduled jobs found for this user'})
            }

        item = response['Items'][0]
        schedule_time = item.get('schedule_time')
        status = item.get('status')

        return {
            'statusCode': 200,
            'body': json.dumps({
                'schedule_time': schedule_time,
                'status': status
            })
        }

    except Exception as e:
        logger.error(f"Error: {str(e)}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }

File: getstatus_lambda/cpp/lambda_function.cpp
#include <aws/core/Aws.h>
#include <aws/lambda-runtime/runtime.h>
#include <aws/dynamodb/DynamoDBClient.h>
#include <aws/dynamodb/model/QueryRequest.h>
#include <aws/dynamodb/model/AttributeValue.h>
#include <aws/core/utils/json/JsonSerializer.h>
#include <aws/core/utils/logging/LogLevel.h>
#include <aws/core/utils/logging/ConsoleLogSystem.h>
#include <aws/core/utils/logging/LogMacros.h>
#include <iostream>
#include <string>
#include <cstdlib>

using namespace aws::lambda_runtime;
using namespace Aws::DynamoDB;
using namespace Aws::DynamoDB::Model;
using namespace Aws::Utils::Json;
using namespace Aws::Utils::Logging;

static invocation_response my_handler(invocation_request const& request)
{
    std::cout << "Lambda handler started" << std::endl;
    std::cout << "Received event: " << request.payload << std::endl;
    JsonValue event(request.payload);
    if (!event.WasParseSuccessful()) {
        std::cerr << "Failed to parse JSON" << std::endl;
        JsonValue fullResponse;
        fullResponse.WithInteger("statusCode", 400);
        fullResponse.WithString("body", JsonValue().WithString("error", "Invalid JSON").View().WriteReadable());
        return invocation_response::success(fullResponse.View().WriteReadable(), "application/json");
    }
    std::cout << "Event JSON parsed successfully" << std::endl;

    JsonView eventView = event.View();
    auto queryParams = eventView.GetObject("queryStringParameters");
    if (!queryParams.IsObject()) {
        std::cerr << "Missing queryStringParameters" << std::endl;
        JsonValue fullResponse;
        fullResponse.WithInteger("statusCode", 400);
        fullResponse.WithString("body", JsonValue().WithString("error", "Missing queryStringParameters").View().WriteReadable());
        return invocation_response::success(fullResponse.View().WriteReadable(), "application/json");
    }
    std::cout << "queryStringParameters found" << std::endl;

    std::string user_id = queryParams.GetString("user_id");
    if (user_id.empty()) {
        std::cerr << "user_id is required" << std::endl;
        JsonValue fullResponse;
        fullResponse.WithInteger("statusCode", 400);
        fullResponse.WithString("body", JsonValue().WithString("error", "user_id is required").View().WriteReadable());
        return invocation_response::success(fullResponse.View().WriteReadable(), "application/json");
    }
    std::cout << "Extracted user_id: " << user_id << std::endl;

    const char* region_env = std::getenv("AWS_APP_REGION");
    std::string region = region_env ? region_env : "eu-north-1";
    std::cout << "Using region: " << region << std::endl;

    const char* table_env = std::getenv("DYNAMODB_TABLE");
    if (!table_env) {
        std::cerr << "DYNAMODB_TABLE environment variable not set" << std::endl;
        JsonValue fullResponse;
        fullResponse.WithInteger("statusCode", 500);
        fullResponse.WithString("body", JsonValue().WithString("error", "DYNAMODB_TABLE environment variable not set").View().WriteReadable());
        return invocation_response::success(fullResponse.View().WriteReadable(), "application/json");
    }

    std::string table_name = table_env;
    std::cout << "Using table: " << table_name << std::endl;

    Aws::Client::ClientConfiguration clientConfig;
    clientConfig.region = region;
    DynamoDBClient dynamoClient(clientConfig);
    std::cout << "DynamoDB client initialized" << std::endl;

    QueryRequest queryRequest;
    queryRequest.SetTableName(table_name);
    Aws::String keyCondition = "user_id = :user_id";
    queryRequest.SetKeyConditionExpression(keyCondition);
    AttributeValue userIdAttr;
    userIdAttr.SetS(user_id);
    queryRequest.AddExpressionAttributeValues(":user_id", userIdAttr);
    queryRequest.SetScanIndexForward(false);
    queryRequest.SetLimit(1);

    std::cout << "Querying DynamoDB for user_id: " << user_id << std::endl;
    auto queryOutcome = dynamoClient.Query(queryRequest);
    if (!queryOutcome.IsSuccess()) {
        std::cerr << "DynamoDB query failed: " << queryOutcome.GetError().GetMessage() << std::endl;
        JsonValue fullResponse;
        fullResponse.WithInteger("statusCode", 500);
        fullResponse.WithString("body", JsonValue().WithString("error", "Internal server error").View().WriteReadable());
        return invocation_response::success(fullResponse.View().WriteReadable(), "application/json");
    }
    std::cout << "DynamoDB query successful" << std::endl;

    auto& items = queryOutcome.GetResult().GetItems();
    if (items.empty()) {
        std::cout << "No items found for user_id: " << user_id << std::endl;
        JsonValue fullResponse;
        fullResponse.WithInteger("statusCode", 404);
        fullResponse.WithString("body", JsonValue().WithString("error", "No scheduled jobs found for this user").View().WriteReadable());
        return invocation_response::success(fullResponse.View().WriteReadable(), "application/json");
    }
    std::cout << "Found " << items.size() << " item(s) for user_id: " << user_id << std::endl;

    auto& item = items[0];
    JsonValue body;
    if (item.count("job_id")) {
        body.WithString("job_id", item.at("job_id").GetS());
    }
    if (item.count("schedule_time")) {
        body.WithString("schedule_time", item.at("schedule_time").GetS());
    }
    if (item.count("status")) {
        body.WithString("status", item.at("status").GetS());
    }
    std::cout << "Response body prepared" << std::endl;

    std::cout << "Returning success for user_id: " << user_id << std::endl;
    JsonValue fullResponse;
    fullResponse.WithInteger("statusCode", 200);
    fullResponse.WithString("body", body.View().WriteReadable());
    std::cout << "Lambda handler completed successfully" << std::endl;
    return invocation_response::success(fullResponse.View().WriteReadable(), "application/json");
}

int main()
{
    Aws::Utils::Logging::InitializeAWSLogging(Aws::MakeShared<Aws::Utils::Logging::ConsoleLogSystem>("lambda", Aws::Utils::Logging::LogLevel::Info));
    Aws::SDKOptions options;
    options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Info;
    Aws::InitAPI(options);
    {
        run_handler(my_handler);
    }
    Aws::ShutdownAPI(options);
    Aws::Utils::Logging::ShutdownAWSLogging();
    return 0;
}
